---
title: "RNAseq analysis"
author: "Eduardo González-Grandío"
date: "June 2nd, 2020"
output: github_document
---
Data preprocessing (done remotely in usegalaxy.org)

1- Trim low quality bases (Trimmomatic)

2- Align paired-end reads (HISAT2)

3- Remove reads that do not map, map to several regions or are unpaired (Filter BAM)

4- Count reads per annotated gene (FeatureCounts): outputs "Feature_count.txt" and "feature_length.tabular" files

5- Compare expression between samples (DESeq2):

Based on https://ucdavis-bioinformatics-training.github.io/2019_March_UCSF_mRNAseq_Workshop/

Load required libraries:
```{r eval = FALSE, message = FALSE}
library(edgeR)
library(biomaRt)
library(data.table)
library(org.At.tair.db)
library(dendextend)
library(WebGestaltR)
library(ComplexHeatmap)
library(pals)
library(ggsci)
library(tidyverse)
library(splitstackshape)
library(stringr)
library(fgsea)
library(gridtext)
```

Calculate DGE object (d) and calculate RPKM (outputs "RPKM.txt" file):

```{r}
library(edgeR)
input="./data/Feature_count.txt"
feature.length="./data/Feature_length.txt"

counts = read.delim(input, row.names = 1) #read output from FeatureCounts
GeneLength = (read.delim(feature.length, row.names = 1)) #read file with length of genes to calculate RPKM, Make sure they are in the same order as the counts file

snames = colnames(counts)
treatment = as.factor(substr(snames, 1, 2))
treatment = factor(treatment, levels=c("NT","WT","RT","DT"))

#Create Differential Gene Expression List Object (DGEList) object
d0 = DGEList(counts = counts, 
             genes = data.frame(Length=GeneLength),
             group = treatment)

d0 = calcNormFactors(d0) #Calculate normalization factors based on library size

RPKM = rpkm(d0) #Calculate reads per kilobase per million reads using normalization factors

RPKM_output = RPKM %>% as.data.frame %>% rownames_to_column("GeneID")
  
write.table(RPKM_output, file = paste0("./output/RPKM.txt"), row.names=F, sep="\t", quote=F)

d0 = d0[filterByExpr(d0), ,keep.lib.sizes=FALSE] # remove genes with low counts and recalculate library size

d0 = calcNormFactors(d0) #Re-calculate normalization factors after removing genes with low amount of reads

```